File: app/src/main/java/aleksandrkim/yandextestprep/Db/Note.java
package aleksandrkim.yandextestprep.Db;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.UUID;

import io.realm.RealmObject;
import io.realm.annotations.PrimaryKey;
import io.realm.annotations.Required;

/**
 * Created by Aleksandr Kim on 07 Apr, 2018 11:27 PM for YandexTestPrep
 */

public class Note extends RealmObject {

    @PrimaryKey
    @Required
    private String id;

    private String title;
    private String content;
    private int color;
    private Date dateCreated;

    public Note() {
        this.id = UUID.randomUUID().toString();
        this.dateCreated = Calendar.getInstance().getTime();
    }

    public Note(String title, String content, int color) {
        this.id = UUID.randomUUID().toString();
        this.dateCreated = Calendar.getInstance().getTime();
        this.title = title;
        this.content = content;
        this.color = color;
    }

    public Note(Note other) {
        this.id = UUID.randomUUID().toString();
        this.dateCreated = Calendar.getInstance().getTime();
        this.title = other.title;
        this.content = other.content;
        this.color = other.color;
    }

    public String getId() {
        return id;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getContent() {
        return content;
    }

    public void setContent(String content) {
        this.content = content;
    }

    public int getColor() {
        return color;
    }

    public void setColor(int color) {
        this.color = color;
    }

    public Date getDateCreated() {
        return dateCreated;
    }

    public String getDateCreatedString() {
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat("dd MMM", new Locale("Ru"));
        return simpleDateFormat.format(dateCreated);
    }

    public void setDateCreated(Date dateCreated) {
        this.dateCreated = dateCreated;
    }
}

File: app/src/main/java/aleksandrkim/MinimalNotes/Db/Note.kt
package aleksandrkim.MinimalNotes.Db

import android.arch.persistence.room.Entity
import android.arch.persistence.room.Ignore
import android.arch.persistence.room.Index
import android.arch.persistence.room.PrimaryKey
import android.support.v7.util.DiffUtil
import java.text.SimpleDateFormat
import java.util.*

/**
 * Created by Aleksandr Kim on 11 Apr, 2018 6:27 PM for ArchComponentsTest
 */

@Entity(tableName = Note.TABLE_NAME,
        indices = [Index("id"), Index(value = ["title", "body"], name = "search_text")])
data class Note(var title: String = "",
                var body: String = "",
                var color: Int = -1,
                var createdTime: Long = Calendar.getInstance().timeInMillis,
                @PrimaryKey(autoGenerate = true)
                var id: Int = 0) {

    @Ignore
    private val simpleDateFormat = SimpleDateFormat("dd MMM", Locale("Ru"))

    @Ignore
    val createdTimeString: String = simpleDateFormat.format(createdTime)

    fun sameAs(obj: Note): Boolean =
        title == obj.title && body == obj.body && color == obj.color && createdTime == obj.createdTime

    fun isBlank() = title.isBlank() && body.isBlank()

    companion object {
        const val TABLE_NAME = "notes"

        val DIFF_ITEM_CALLBACK: DiffUtil.ItemCallback<Note> = object : DiffUtil.ItemCallback<Note>() {
            override fun areItemsTheSame(oldItem: Note, newItem: Note): Boolean = oldItem.id == newItem.id

            override fun areContentsTheSame(oldItem: Note, newItem: Note): Boolean = oldItem.equals(newItem)
        }
    }
}
